main.cpp

```diff
@@ -6,134 +6,714 @@
+DoubleType::~DoubleType() { delete value; }
+
+DoubleType& DoubleType::add(const double& rhs)
+{
+    *value += rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::add(const FloatType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value += static_cast<double>(*rhs.value);
+    return *this;
+}
+
+DoubleType& DoubleType::add(const DoubleType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value += *rhs.value;
+    return *this;
+}
+
+DoubleType& DoubleType::add(const IntType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value += static_cast<double>(*rhs.value);
+    return *this;
+}
+
+DoubleType& DoubleType::subtract(const double& rhs)
+{
+    *value -= rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::subtract(const FloatType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value -= static_cast<double>(*rhs.value);
+    return *this;
+}
+
+DoubleType& DoubleType::subtract(const DoubleType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value -= *rhs.value;
+    return *this;
+}
+
+DoubleType& DoubleType::subtract(const IntType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value -= static_cast<double>(*rhs.value);
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(const double& rhs)
+{
+    *value *= rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(const FloatType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value *= static_cast<double>(*rhs.value);
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(const DoubleType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value *= *rhs.value;
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(const IntType& rhs)
+{
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    *value *= static_cast<double>(*rhs.value);
+    return *this;
+}
+
+DoubleType& DoubleType::divide(const double& rhs)
+{
+    if (rhs == 0.0)
+    {
+        std::cout << "divide by zero\n";
+        return *this;
+    }
+    *value /= rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::divide(const FloatType& rhs)
 {
-    FloatType ft;
-    DoubleType dt;
-    IntType it;
+    if (rhs.value == nullptr)
+    {
+        std::cout << "value null ptr\n";
+        return *this;
+    }
+    if (*rhs.value == 0.f)
+    {
+        std::cout << "divide by zero\n";
+        return *this;
```

Comment:
- [ ] floating point division by 0 is not an error.  you don't need to return yet.